crates/bevy_ecs/src/system/function_system.rs

```diff
@@ -489,3 +507,21 @@ macro_rules! impl_system_function {
 }
 
 all_tuples!(impl_system_function, 0, 16, F);
+
+pub trait AsSystemLabel<Marker> {
+    fn as_system_label(&self) -> Box<dyn SystemLabel>;
+}
+
+impl<In, Out, Param: SystemParam, Marker, T: SystemParamFunction<In, Out, Param, Marker>>
+    AsSystemLabel<(In, Out, Param, Marker)> for T
+{
+    fn as_system_label(&self) -> Box<dyn SystemLabel> {
+        Box::new(SystemTypeIdLabel(self.type_id()))
+    }
+}
+
+impl<T: SystemLabel> AsSystemLabel<()> for T {
+    fn as_system_label(&self) -> Box<dyn SystemLabel> {
+        self.dyn_clone()
```

Comment:
The need to clone here is an unfortunate 'regression'

Is there any downside to just making `AsSystemLabel` consume self?

Reply:
Consuming `self` would make this unusable for closure systems, as retrieving the label would consume the IntoSystem impl (aka the closure).

Reply:
Note that this doesn't add extra boxing (this replaces the manual boxing that we did in the past). Given that basically all system labels are enums, typeids, or ZSTs, the extra clone probably wouldn't even be measurable, even for many thousands of systems.

Reply:
It's already hugely unergonomic for closure systems though. Either you create the 'same' closure twice:

```rust
fn my_closure_system() -> ... {||{}}
app.with_system(my_closure_system()).with_system(other.after(my_closure_system()))
```
Or you store it in a local variable:
```rust
let closure = || {};
with_system(other.after(&closure)).with_system(closure) // has to be this order.
```

But yeah, perhaps being able to use the second pattern is worthwhile.

Reply:
IMO this is so challenging (and niche) for closure systems that we shouldn't go out of our way to support it. Providing an explicit label is a nice clear workaround.

Reply:
This seems like a really reasonable use case / pattern for closure systems. Given that closure systems are now "the" way to pass in-scope configuration into a system, and closures will also exist in that scope, I think its reasonable to support ordering without needing to define custom labels. At the very least, I think this should be possible (and it currently is):

```rust
let closure = || {};
app
  .add_system(foo.after(closure.as_system_label())))
  .add_system(closure)
```

But I'm also open to making @DJMcNab's second suggestion work.